packages/api/lambdas/sf-event-sqs-to-db-records.js

```diff
@@ -82,50 +119,124 @@ const shouldWriteExecutionToRDS = async (cumulusMessage, knex) => {
+      status: getWorkflowStatus(cumulusMessage),
+      collectionCumulusId: collection.cumulusId,
+      providerCumulusId: provider.cumulusId,
     });
-  } catch (error) {
-    log.error(`Failed to write execution records for ${executionArn}`, error);
-    throw error;
+
+const savePdr = async ({
+  cumulusMessage,
+  collection,
+  provider,
+  knex,
+  pdrModel = new Pdr(),
+}) => {
+  // If there is no PDR in the message, then there's nothing to do here, which is fine
+  if (!messageHasPdr(cumulusMessage)) {
+    return true;
+  }
+  if (!isRecordDefined(collection)) {
+    throw new Error(`Collection reference is required for a PDR, got ${collection}`);
   }
+  if (!isRecordDefined(provider)) {
+    throw new Error(`Provider reference is required for a PDR, got ${provider}`);
+  }
+  return knex.transaction(async (trx) => {
+    await savePdrViaTransaction({ cumulusMessage, collection, provider, trx });
+    return pdrModel.storePdrFromCumulusMessage(cumulusMessage);
+  });
 };
 
-const savePdrToDb = async (cumulusMessage) => {
-  const pdrModel = new Pdr();
+const saveGranulesToDb = async (cumulusMessage) => {
+  const granuleModel = new Granule();
+
   try {
-    await pdrModel.storePdrFromCumulusMessage(cumulusMessage);
+    await granuleModel.storeGranulesFromCumulusMessage(cumulusMessage);
   } catch (error) {
     const executionArn = getMessageExecutionArn(cumulusMessage);
-    log.fatal(`Failed to create/update PDR database record for execution ${executionArn}: ${error.message}`);
+    log.fatal(`Failed to create/update granule records for execution ${executionArn}: ${error.message}`);
     throw error;
   }
 };
 
-const saveGranulesToDb = async (cumulusMessage) => {
+const saveRecordsToDynamoDb = async (cumulusMessage) => {
+  const executionModel = new Execution();
+  const pdrModel = new Pdr();
   const granuleModel = new Granule();
 
+  const results = await Promise.allSettled([
+    executionModel.storeExecutionFromCumulusMessage(cumulusMessage),
+    pdrModel.storePdrFromCumulusMessage(cumulusMessage),
+    granuleModel.storeGranulesFromCumulusMessage(cumulusMessage),
+  ]);
+  const failures = results.filter((result) => result.status === 'rejected');
+  if (failures.length > 0) {
+    const allFailures = failures.map((failure) => failure.reason);
+    log.error(allFailures.join(' '));
+    throw new Error('Failed writing some records to Dynamo');
+  }
+  return results;
+};
+
+const saveRecords = async (cumulusMessage, knex) => {
+  const executionArn = getMessageExecutionArn(cumulusMessage);
+
+  const collection = await getMessageCollection(cumulusMessage, knex);
+  const isExecutionRDSWriteEnabled = await shouldWriteExecutionToRDS(
+    cumulusMessage,
+    collection,
+    knex
+  );
+
+  // If execution is not written to RDS, then PDRs/granules which reference
+  // execution should not be written to RDS either
+  if (!isExecutionRDSWriteEnabled) {
+    return saveRecordsToDynamoDb(cumulusMessage);
```

Comment:
The logic has been refactored here because I realized: if the execution isn't written to RDS, then neither PDRs or granules should be either, since they both would want to refer to the execution in RDS. Does that seem correct?

Reply:
Yes... but is that true for granules as well?    We can save that for the next PR.

Reply:
I'm fairly certain it would also be true for granules? Granules refer to an execution, so if the execution wasn't written to RDS, those granule RDS writes would fail so we shouldn't attempt any RDS write for them. Does that sound right?

Reply:
Probably.  The situation I'm still worried about is a situation where we notify CMR about a granule, but fail to have a granule record in Core.     From a data model/db perspective, obviously you want a complete record, but I have some disquiet about what users are going to want in this regard.

  Right now in Dynamo if an execution write fails the granule *may or may not*  write( thanks promise.all), so that behavior is suboptimal anyway. 

That concern is *probably* best relegated to the "what do we do about DLQ messages" and/or "reconciliation report handling" feature/discussion/ether though.

Reply:
The difference is that in Dynamo, we're not making the granule write dependent on the execution write. The granule dynamo write may still fail for _reasons_ though.

But yeah, agree that this is really a question of how to handle/surface errors produced by this Lambda, which I think is outside this ticket